text/0517-io-os-reform.md

```diff
@@ -708,7 +712,129 @@ throughout IO, we can go on to explore the modules in detail.
 ### `std::fs`
 [std::fs]: #stdfs
 
-> To be added in a follow-up PR.
+The `fs` module will provide most of the functionality it does today,
+but with a stronger cross-platform orientation.
+
+Note that all path-consuming functions will now take an
+`AsPath`-bounded parameter for ergonomic reasons (this will allow
+passing in Rust strings and literals directly, for example).
+
+#### Free functions
+[Free functions]: #free-functions
+
+**Files**:
+
+* `copy`. Take `AsPath` bound.
+* `rename`. Take `AsPath` bound.
+* `remove_file` (renamed from `unlink`). Take `AsPath` bound.
+
+* `file_attr` (renamed from `stat`). Take `AsPath` bound. Yield a new
+  struct, `FileAttr`, with no public fields, but `size`, `kind` and
+  `perm` accessors. The various `os::platform` modules will offer
+  extension methods on this structure.
+
+* `set_perm` (renamed from `chmod`). Take `AsPath` bound, and a
+  `FilePermissions` value. The `FilePermissions` type will be revamped
+  as a struct with private implementation; see below.
+
+**Directories**:
+
+* `make_dir` (renamed from `mkdir`). Take `AsPath` bound.
+* `make_dir_all` (renamed from `mkdir_recursive`). Take `AsPath` bound.
+* `read_dir` (renamed from `readdir`). Take `AsPath` bound. Yield a
+  newtypes iterator, which yields a new type `DirEntry` which has an
+  accessor for `Path`, but will eventually provide other information
+  as well (possibly via platform-specific extensions).
+* `remove_dir` (renamed from `rmdir`). Take `AsPath` bound.
+* `remove_dir_all` (renamed from `rmdir_recursive`). Take
+  `AsPath` bound.
+* `walk_dir`. Take `AsPath` bound. Yield an iterator over `IoResult<DirEntry>`.
+
+**Links**:
+
+* `hard_link` (renamed from `link`). Take `AsPath` bound.
+* `sym_link` (renamed from `symlink`). Take `AsPath` bound.
+* `read_link` (renamed form `readlink`). Take `AsPath` bound.
+
+#### Files
+[Files]: #files
+
+The `File` type will largely stay as it is today, except that it will
+use the `AsPath` bound everywhere.
+
+The `stat` method will be renamed to `attr`, yield a `FileAttr`, and
+take `&self`.
+
+The `fsync` method will be renamed to `flush`, and `datasync` will be
+renamed to `flush_data`. (Although the latter is not available on
+Windows, it can be considered an optimization for `flush` and on
+Windows behave identically to `flush`, just as it does on some Unix
+filesystems.)
+
+The `path` method wil remain `#[unstable]`, as we do not yet want to
+commit to its API.
+
+The `open_mode` function will take an `OpenOptions` struct, which will
+encompass today's `FileMode` and `FileAccess` and support a
+builder-style API.
+
```

Comment:
One method I think we forgot to consider is `truncate` on files today. Right now the function matches [Unix semantics](http://linux.die.net/man/2/ftruncate) where it basically acts as a `set_len` function. It will shorten the file or extend it (filling in with 0s) if necessary.

This function, however doesn't exist on Windows to my knowledge (cc @retep998, I may be missing something). Currently we're using [`SetEndOfFile`](https://msdn.microsoft.com/en-us/library/windows/desktop/aa365531%28v=vs.85%29.aspx) which operates slightly different than the Unix variant. On Windows you must first `seek` to a position and *then* "truncate" the file, which is to say that Windows will simply set the length of the file on disk to the current pointer in the file.

Unfortunately, the current semantics may lead to some loss on windows. Our "emulation" of `ftruncate` on Windows involves seeking the file to the desired length, calling `SetEndOfFile`, and then seeking back to where the file used to be. The "seek back" operation, however, may fail, which would mean the operation actually completed successfully but we're forced to still return an error.

I think I would personally recommend matching Windows semantics rather than unix semantics in this case. We would drop the `u64` argument to the function, call `SetEndOfFile` on Windows, and unix would be a combination of `lseek` (to learn the position) followed by `ftruncate`.